src/common/snippets/src/pass/common_optimizations.cpp

```diff
@@ -19,6 +19,244 @@ namespace ov {
 namespace snippets {
 namespace pass {
 
+bool CommonOptimizations::canBeParallelismWAOptimized(const std::shared_ptr<const ov::Node>& node, size_t num_threads) {
+    if (!ov::is_type<ov::op::v0::MatMul>(node))
+        return false;
+    // It's needed only for 3D MHA patterns
+    const auto mm_shape = node->get_shape();
+    if (mm_shape.size() != 3)
+        return false;
+    const auto current_parallel_work_amount =
+        std::accumulate(mm_shape.rbegin() + 2, mm_shape.rend(), size_t(1), std::multiplies<size_t>());
+    const auto dim_M = *(mm_shape.rbegin() + 1);
+    return (current_parallel_work_amount < num_threads) && (current_parallel_work_amount * dim_M >= num_threads);
+}
+
+void CommonOptimizations::SplitDimensionM(const std::shared_ptr<ov::snippets::op::Subgraph>& subgraph, size_t num_threads) {
+    // To increase parallelism work in 3D cases for MHA pattern,
+    // we split 1st dimension (starting from 0th) into 2 new dimensions to get 4D Shapes where
+    // - 0th and 1st dimensions are used in parallel scheduling,
+    // - 2nd and 3rd dimensions are used in kernel
+    // Note: 3D Patterns don't contain Transpose inside so the reshaping is valid
+
+    // It's needed only for MHA patterns. Need to add support for common patterns
+    if (!subgraph->has_domain_sensitive_ops())
+        return;
+
+    const auto& body = subgraph->body_ptr();
+    const auto& parameters = body->get_parameters();
+    // [107806]: If count of Parameters isn't equal to Subgraph inputs (it's possible case in general),
+    //           we cannot garantee correct extraction since we don't have correct connections between body I/O and Subgraph I/O.
+    OPENVINO_ASSERT(parameters.size() == subgraph->input_values().size(),
+                    "Failed to extract unsupported transposes: the count of Parameters isn't equal to Subgraph inputs");
+
+    // Need to find MatMul0 and check output shape
+    const auto& ops = body->get_ordered_ops();
+    const auto mm_it = std::find_if(ops.begin(), ops.end(),
+                                    [](const std::shared_ptr<ov::Node>& node){ return ov::is_type<ov::op::v0::MatMul>(node); });
+    if (mm_it == ops.end())
+        return;
+
+    const auto matmul0 = ov::as_type_ptr<ov::op::v0::MatMul>(*mm_it);
+    if (!matmul0 || !canBeParallelismWAOptimized(matmul0, num_threads))
+        return;
+
+    auto get_dim_M = [](const ov::Shape& shape) {
+        return *(shape.rbegin() + 1);
+    };
+
+    const auto mm_shape = matmul0->get_shape();
+    const auto optimal_parallelism_work_amount = static_cast<size_t>(num_threads);
+    const auto optimal_m_dim = 64;  // heuristic
```

Comment:
I agree. It was just for testing. But seems like this value depends on blocking parameters which are implemented only for one backends. Probable when we will move this optimization to Subgraph (like dimension collapsion), we add interface method `get_optimal_M_dim` which will be implemented by each backend (generator?).
Thus, I think we can so but then update this logic. What do you think? Maybe do you see another solution at the moment? It'd be really great if you share with me since unfortunately I cannot сome up with another more common solution yet